Competitive Programming : Edit Distance Problem Dynamic Programming (Documentation + Code)

pranshukas opened the pull request:
## Description 📜

Fixes #1224 
Competitive Programming: Edit Distance Problem (Documentation + Code) #1224
Added Documentation for Edit Distance Problem, both Recursive + Memoization (Top Down) and Iterative (Bottom Up) Approach of Dynamic Programming.

<hr>

## Type of change 📝

<!----Please delete options that are not relevant.And in order to tick the check box just but x inside them for example [x] like this----->

- [x] Documentation (Content Creation in the form of codes or tutorials)

<hr>

## Domain of Contribution 📊

<!----Please delete options that are not relevant.And in order to tick the check box just but x inside them for example [x] like this----->

- [x] Competitive Programming

<hr>

## Checklist ✅

<!----Please delete options that are not relevant.And in order to tick the check box just but x inside them for example [x] like this----->

- [x] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [x] I have performed a self-review of my own code or work.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generates no new warnings.
- [x] I'm GWOC'21 contributor

<hr>

pranshukas commented:
Please review my PR and revert in case of any changes need to be made. 
Any Suggestions/Feedbacks or Scope for improvements are Welcomed !😊😊

atarax665 reviewed (changes requested):
- Make a folder with the name Edit Distance inside DP folder and add this documentation there. -- Also, rename the file to `edit_distance_documentation.md`.
- In the code part please keep the comments above the line of code, not beside the lines, this would make the code look cleaner.

Nice work 👍

pranshukas commented:
Thank You so much I made the changes you suggested, you can view. Can You tell me one thing? I just made 2 commits i.e, added folder Edit Distance and updated edit_distance_documentation.md and fetched the upstream but its showing 319 commits. 
Did I made something wrong.

atarax665 commented on Competitive_Programming/C++/DP/Edit Distance/edit_distance_documentation.md:
```diff
@@ -0,0 +1,149 @@
+2. Remove a character
+3. Replace a character
+
+Cost of all the above operations is same. Find the minimum number of operations that are required to convert ‘str1’ into ‘str2’.  
```
```suggestion
The cost of all the above operations is the same. Find the minimum number of operations that are required to convert ‘str1’ into ‘str2’.  
```

atarax665 commented on Competitive_Programming/C++/DP/Edit Distance/edit_distance_documentation.md:
```diff
@@ -0,0 +1,149 @@
+
+## Naive Approach 🧠
+
+A naive brute force idea is to traverse on the strings and process each character one by one. Let traverse both the strings from right to left. 
```
```suggestion
A naive brute force idea is to traverse the strings and process each character one by one. Lets traverse both the strings from right to left. 
```

atarax665 commented on Competitive_Programming/C++/DP/Edit Distance/edit_distance_documentation.md:
```diff
@@ -0,0 +1,149 @@
+n = Lenght of String 2
+There can be two possiblity 
+
+**Case 1**: Last Characters of both the Strings are same. Nothing to Do, so cost=0 move to next index, i.e, Recur for (n-1,m-1)
```
```suggestion
**Case 1**: Last Characters of both the Strings are the same. Nothing to do, so cost=0 move to next index, i.e, Recur for (n-1,m-1)
```

atarax665 commented on Competitive_Programming/C++/DP/Edit Distance/edit_distance_documentation.md:
```diff
@@ -0,0 +1,149 @@
+
+![image](https://user-images.githubusercontent.com/58984074/134164390-56828bac-3a4f-4fbd-9b9b-3211fdf02dd1.png)
+
+It can be concluded that are lots of overlapping subproblems that are solved again and again. Example f(2,2) is calculated many times. Hence we can store the intermediate results of computations and use them further if required. 
```
```suggestion
It can be concluded that there are lots of overlapping subproblems that are solved again and again. Example f(2,2) is calculated many times. Hence we can store the intermediate results of computations and use them further if required. 
```

atarax665 commented on Competitive_Programming/C++/DP/Edit Distance/edit_distance_documentation.md:
```diff
@@ -0,0 +1,149 @@
+
+## 💡 Implementation of Recursion + Memoization (Top Down Approach):
+
+The above Naive Recursive approach above can be optimized using memoization. A 2-D Vector can be used to store the intermediate results and for every state f(n,m) it can be checked whether it has been calculated before. 
```
```suggestion
The above Naive Recursive approach above can be optimized using memoization. A 2-D vector can be used to store the intermediate results and for every state f(n,m) it can be checked whether it has been calculated before. 
```

atarax665 reviewed (approved):
LGTM 👍
